contracts/auction/ClockAuction.sol

```diff
@@ -121,30 +152,90 @@ contract ClockAuction is AuctionRelated {
+        // at least bidWaitingTime after last bidder's bid moment,
+        // and no one else has bidden during this bidWaitingTime,
+        // then any one can claim this token(land) for lastBidder.
+        require(now >= auction.lastBidStartAt + bidWaitingTime,
+        "this auction has not finished yet, try again later");
+        //prevent re-entry attack
+        _removeAuction(_tokenId);
 
+        nonFungibleContract.safeTransferFrom(this, auction.lastBidder, _tokenId);
+
+        emit AuctionSuccessful(_tokenId, auction.lastRecord, auction.lastBidder);
+    }
 
     // @dev bid with RING. Computes the price and transfers winnings.
     function _bidWithRING(address _from, uint256 _tokenId, uint256 _valueInRING) internal returns (uint256){
+        // Get a reference to the auction struct
         Auction storage auction = tokenIdToAuction[_tokenId];
         require(_isOnAuction(auction));
-        address seller = auction.seller;
-        // current price in RING
+
+        require(now <= auction.lastBidStartAt + bidWaitingTime);
+
+        // Check that the incoming bid is higher than the current price
         uint priceInRING = _currentPriceInRING(auction);
-        require (_valueInRING >= priceInRING);
-        _removeAuction(_tokenId);
+        require (_valueInRING >= priceInRING,
+            "your offer is lower than the current price, try again with a higher one.");
+
+        // TODO: record last bidder's info
+        // last bidder's info
+        address lastBidder;
+        // last bidder's price
+        uint lastRecord;
+
+        if (lastBidder != 0x0) {
+            lastBidder = auction.lastBidder;
+            lastRecord = uint256(auction.lastRecord);
+        }
 
-        if (priceInRING > 0) {
-            uint256 auctioneerCutInRING = _computeCut(priceInRING);
-            uint256 sellerProceedsInRING = priceInRING - auctioneerCutInRING;
+        // TODO: modify bid-related member variables
+        // modify bid-related member variables
+        uint bidMoment = now;
+        auction.lastBidder = _from;
+        auction.lastRecord = uint128(priceInRING);
+        auction.lastBidStartAt = bidMoment;
+
+        // Grab a reference to the seller before the auction struct
+        // gets deleted.
+        address seller = auction.seller;
+
+        if (lastBidder == 0x0 && priceInRING > 0) {
+            //  Calculate the auctioneer's cut.
+            // (NOTE: _computeCut() is guaranteed to return a
+            //  value <= price, so this subtraction can't go negative.)
+            uint256 sellerProceedsInRING = priceInRING - _computeCut(priceInRING);
+            // transfer to the seller
             RING.transfer(seller, sellerProceedsInRING);
-            emit AuctionSuccessful(_tokenId, priceInRING, _from);
         }
 
+        //  not the first bid
+        if (lastRecord > 0 && lastBidder != 0x0) {
+            uint extraForEach = (priceInRING - lastRecord) / 2;
```

Comment:
make sure priceInRING is large than lastRecord, and use SafeMath to avoid overflow and review efforts.